Mathlib/Data/Matrix/Composition.lean

```diff
@@ -0,0 +1,132 @@
+/-
+Copyright (c) 2024 Yunzhou Xie. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin Buzzard, Yunzhou Xie
+-/
+import Mathlib.Data.Matrix.Basic
+
+/-!
+# Composition of matrices
+This file shows that Mₙ(Mₘ(R)) ≃ Mₙₘ(R), Mₙ(Mₘ(R)) ≃ Mₘ(Mₙ(R)), Mn(Rᵒᵖ) ≃ₐ[K] Mₙ(R)ᵒᵖ
+and also different levels of equivalence when R is an AddCommMonoid,
+Semiring, and Algebra over a CommSemiring K.
+-/
+namespace Matrix
+
+open BigOperators
+
+variable  (I J K L R : Type*)
+
+/-- Mₙ(Mₘ(R)) ≃ Mₙₘ(R) -/
+def comp : Matrix I J (Matrix K L R) ≃ Matrix (I × K) (J × L) R where
+  toFun m ik jl := m ik.1 jl.1 ik.2 jl.2
+  invFun n i j k l := n (i, k) (j, l)
+  left_inv _ := rfl
+  right_inv _ := rfl
+
+/-- Mₙ(Mₘ(R)) ≃ Mₘ(Mₙ(R)) -/
+def swap : Matrix (I × J) (K × L) R ≃ Matrix (J × I) (L × K) R where
+  toFun m ji kl := m (ji.2, ji.1) (kl.2, kl.1)
+  invFun n ij kl := n (ij.2, ij.1) (kl.2, kl.1)
+  left_inv _ := rfl
+  right_inv _ := rfl
+
+section AddCommMonoid
+variable [AddCommMonoid R]
+
+/-- Mₙ(Mₘ(R)) ≃+ Mₙₘ(R) -/
+def comp_addHom : Matrix I J (Matrix K L R) ≃+ Matrix (I × K) (J × L) R :=
+{ Matrix.comp I J K L R with
+  map_add' := fun _ _ ↦ rfl
+}
+
+/-- Mₙ(Mₘ(R)) ≃+ Mₘ(Mₙ(R)) -/
+def swap_addHom : Matrix (I × J) (K × L) R ≃+ Matrix (J × I) (L × K) R :=
+{ Matrix.swap I J K L R with
+  map_add' := fun _ _ ↦ rfl
+}
+
+end AddCommMonoid
+
+section Semiring
+variable [Semiring R][Fintype I][Fintype J] [DecidableEq I] [DecidableEq J]
+
+/-- Mₙ(Mₘ(R)) ≃+* Mₙₘ(R) -/
+def comp_ringHom : Matrix I I (Matrix J J R) ≃+* Matrix (I × J) (I × J) R :=
```

Comment:
I think the naming convention should be `compRingHom`

Reply:
Any specific reasons? :(

Reply:
ok done